// json-web-token/dist/index.mjs
import { createHmac, createSign, timingSafeEqual, createVerify } from 'crypto';

// src/index.ts
var algorithms = {
  HS256: { hash: "sha256", type: "hmac" },
  HS384: { hash: "sha384", type: "hmac" },
  HS512: { hash: "sha512", type: "hmac" },
  RS256: { hash: "RSA-SHA256", type: "sign" }
};
var JWTError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "JWTError";
  }
};
function isPlainObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}
function inferKeyKind(key) {
  const head = Buffer.isBuffer(key) ? key.subarray(0, 11).toString("ascii") : key.trimStart().slice(0, 11);
  return head === "-----BEGIN " ? "asymmetric" : "secret";
}
function algorithmMatchesKey(alg, kind) {
  return alg.type === "hmac" ? kind === "secret" : kind === "asymmetric";
}
function isWrapper(value) {
  return isPlainObject(value) && "payload" in value && value.payload != null;
}
function fail(message) {
  return { error: new JWTError(message), value: null };
}
function failDecode(message) {
  return { error: new JWTError(message), value: null };
}
function getAlgorithms() {
  return Object.keys(algorithms);
}
function encode(key, data, algorithm = "HS256") {
  const defaultHeader = { typ: "JWT", alg: algorithm };
  const payload = isWrapper(data) ? data.payload : data;
  const header = isWrapper(data) && isPlainObject(data.header) ? { ...data.header, ...defaultHeader } : defaultHeader;
  const validationError = encodeValidations(key, payload, algorithm);
  if (validationError) return fail(validationError);
  const algSpec = algorithms[algorithm];
  if (!algorithmMatchesKey(algSpec, inferKeyKind(key))) {
    return fail(
      algSpec.type === "hmac" ? "Algorithm/key mismatch: HMAC algorithm requires a plain secret, not an asymmetric key." : "Algorithm/key mismatch: asymmetric algorithm requires a PEM-encoded key, not a plain secret."
    );
  }
  const headerPart = encodeBase64Url(JSON.stringify(header));
  const payloadPart = encodeBase64Url(JSON.stringify(payload));
  const signingInput = `${headerPart}.${payloadPart}`;
  const signature = sign(algSpec, key, signingInput);
  return { error: null, value: `${signingInput}.${signature}` };
}
function decode(key, token, options = {}) {
  if (!key || !token) return failDecode("The key and token are mandatory!");
  const parts = token.split(".");
  if (parts.length !== 3) {
    return failDecode("The JWT should consist of three parts!");
  }
  const header = parseJSONOrNull(decodeBase64Url(parts[0]));
  const payload = parseJSONOrNull(decodeBase64Url(parts[1]));
  if (!isPlainObject(header)) {
    return failDecode("The algorithm is not supported!");
  }
  const algName = header.alg;
  const algorithm = typeof algName === "string" ? algorithms[algName] : void 0;
  if (!algorithm) return failDecode("The algorithm is not supported!");
  if (options.algorithms && !options.algorithms.includes(algName)) {
    return failDecode("The algorithm is not in the allowlist.");
  }
  if (!algorithmMatchesKey(algorithm, inferKeyKind(key))) {
    return failDecode(
      algorithm.type === "hmac" ? "Algorithm/key mismatch: refusing to verify an HMAC signature with an asymmetric key." : "Algorithm/key mismatch: refusing to verify an asymmetric signature with a plain secret."
    );
  }
  const ok = verify(
    algorithm,
    key,
    `${parts[0]}.${parts[1]}`,
    parts[2]
  );
  return {
    error: ok ? null : new JWTError("Invalid key!"),
    value: payload,
    header
  };
}
function encodeValidations(key, payload, algorithm) {
  if (!key || payload == null) return "The key and payload are mandatory!";
  if (isPlainObject(payload) && Object.keys(payload).length === 0) {
    return "The payload is an empty object!";
  }
  if (!algorithms[algorithm]) {
    return "The algorithm is not supported!";
  }
  return null;
}
function sign(alg, key, input) {
  if (alg.type === "hmac") {
    return createHmac(alg.hash, key).update(input).digest("base64url");
  }
  return createSign(alg.hash).update(input).sign(key, "base64url");
}
function verify(alg, key, input, signature) {
  if (alg.type === "hmac") {
    const expected = sign(alg, key, input);
    const a = Buffer.from(expected, "utf8");
    const b = Buffer.from(signature, "utf8");
    if (a.length !== b.length) return false;
    return timingSafeEqual(a, b);
  }
  return createVerify(alg.hash).update(input).verify(key, signature, "base64url");
}
function encodeBase64Url(input) {
  return Buffer.from(input, "utf8").toString("base64url");
}
function decodeBase64Url(input) {
  return Buffer.from(input, "base64url").toString("utf8");
}
function parseJSONOrNull(input) {
  try {
    return JSON.parse(input);
  } catch {
    return null;
  }
}

export { JWTError, decode, encode, getAlgorithms };
